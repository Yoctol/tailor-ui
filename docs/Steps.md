---
id: steps
title: Steps
---

`Steps` is a navigation bar that guides users through the steps of a task.

## When To Use

When the task is complicated or has a certain sequence in the series of subtasks, we can decompose it into several steps to make things easier.

## Examples

### Basic

```jsx live
<>
  <Steps current={1}>
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1}>
    <Steps.Step title="Finished" />
    <Steps.Step title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</>
```

### Error status

```jsx live
<>
  <Steps current={1}>
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step
      status="error"
      title="In Progress"
      description="This is a description"
    />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1}>
    <Steps.Step title="Finished" />
    <Steps.Step status="error" title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</>
```

### Vertical

```jsx live
<>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" />
    <Steps.Step title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</>
```

### Vertical & Error status

```jsx live
<>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step
      status="error"
      title="In Progress"
      description="This is a description"
    />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" />
    <Steps.Step status="error" title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</>
```

### Switch Step

```jsx live
() => {
  const message = useMessage();
  const [step, setStep] = useState(0);

  const steps = [
    {
      title: 'First',
      content: 'First-content',
    },
    {
      title: 'Second',
      content: 'Second-content',
    },
    {
      title: 'Last',
      content: 'Last-content',
    },
  ];

  return (
    <div>
      <Steps current={step} onCurrentChange={setStep}>
        {steps.map(({ title }) => (
          <Steps.Step
            key={title}
            title={title}
            description="This is a description"
          />
        ))}
      </Steps>

      <Flex
        my="4"
        height="200px"
        alignItems="center"
        justifyContent="center"
        bg="gray300"
        borderRadius="1"
      >
        {steps[step].content}
      </Flex>

      {step < steps.length - 1 && (
        <Button variant="primary" onClick={() => setStep(step + 1)}>
          Next
        </Button>
      )}
      {step === steps.length - 1 && (
        <Button
          variant="primary"
          onClick={() => message.success('Processing complete!')}
        >
          Done
        </Button>
      )}
      {step > 0 && (
        <Button ml="2" onClick={() => setStep(step - 1)}>
          Previous
        </Button>
      )}
    </div>
  );
}
```

## API

### Steps

### Steps.Step
