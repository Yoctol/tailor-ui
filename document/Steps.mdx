---
name: Steps
menu: Components
route: /components/steps
---

import { PropsTable, Playground } from 'docz';
import { Button, Flex, Steps, message } from 'tailor-ui';

# Steps

`Steps` is a navigation bar that guides users through the steps of a task.

## When To Use

When the task is complicated or has a certain sequence in the series of subtasks, we can decompose it into several steps to make things easier.

## Examples

### Basic

<Playground>
  <Steps current={1}>
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1}>
    <Steps.Step title="Finished" />
    <Steps.Step title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Error status

<Playground>
  <Steps current={1}>
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step
      status="error"
      title="In Progress"
      description="This is a description"
    />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1}>
    <Steps.Step title="Finished" />
    <Steps.Step status="error" title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Vertical

<Playground>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" />
    <Steps.Step title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Vertical & Error status

<Playground>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step
      status="error"
      title="In Progress"
      description="This is a description"
    />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br />
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" />
    <Steps.Step status="error" title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Switch Step

<Playground>
  {() => {
    const [step, setStep] = React.useState(0);

    const steps = [{
      title: 'First',
      content: 'First-content',
    }, {
      title: 'Second',
      content: 'Second-content',
    }, {
      title: 'Last',
      content: 'Last-content',
    }];

    return (
      <div>
        <Steps current={step} onCurrentChange={setStep}>
          {steps.map(({ title }) => (
            <Steps.Step key={title} title={title} description="This is a description" />
          ))}
        </Steps>

        <Flex
          my="4"
          height="200px"
          alignItems="center"
          justifyContent="center"
          bg="gray300"
          borderRadius="1"
        >
          {steps[step].content}
        </Flex>

        {step < steps.length - 1 && (
          <Button type="primary" onClick={() => setStep(step + 1)}>Next</Button>
        )}
        {step === steps.length - 1 && (
          <Button
            type="primary"
            onClick={() => message.success('Processing complete!')}
          >
            Done
          </Button>
        )}
        {step > 0 && (
          <Button ml="2" onClick={() => setStep(step - 1)}>
            Previous
          </Button>
        )}
      </div>
    );

}}

</Playground>

## API

### Steps

<PropsTable of={Steps} />

### Steps.Step

<PropsTable of={Steps.Step} />
