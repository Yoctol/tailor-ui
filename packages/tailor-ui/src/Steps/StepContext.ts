import { createContext } from 'react';

import { theme } from '@tailor-ui/theme';

export type Status = 'finish' | 'progress' | 'wait' | 'error';
export type Direction = 'horizontal' | 'vertical';
export type Colors = keyof typeof theme.colors;

const StepContext = createContext<{
  count: number;
  status: Status;
  isLast: boolean;
  tailColor: Colors | 'gray';
  direction: Direction;
  onCurrentChange?: (count: number) => void;
}>({
  count: 0,
  status: 'wait',
  isLast: false,
  tailColor: 'gray',
  direction: 'horizontal',
  onCurrentChange: () => {},
});

export default StepContext;
