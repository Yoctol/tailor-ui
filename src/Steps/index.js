import ErrorIcon from 'react-icons/lib/md/close';
import FinishIcon from 'react-icons/lib/md/done';
import PropTypes from 'prop-types';
import React, { Children, createContext } from 'react';
import styled, { css } from 'styled-components';

import Flex from '../Grid/Flex';
import Icon from '../Icon';

const { Provider, Consumer } = createContext();

const statusType = PropTypes.oneOf(['finish', 'progress', 'wait', 'error']);

const StepsIconItem = styled.div`
  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  margin-right: 8px;
  border: ${p => p.theme.borders.default};
  border-radius: 50%;
  border-color: ${p => {
    switch (p.status) {
      case 'progress':
      case 'finish':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.error;
      case 'wait':
      default:
        return p.theme.colors.gray[5];
    }
  }};
  background-color: ${p =>
    p.status === 'progress' ? p.theme.colors.primary : p.theme.colors.light};
  color: ${p => {
    switch (p.status) {
      case 'progress':
        return p.theme.colors.light;
      case 'finish':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.error;
      case 'wait':
      default:
        return p.theme.colors.gray[5];
    }
  }};
  cursor: pointer;

  ${p =>
    !p.isLast &&
    p.direction === 'vertical' &&
    css`
      &::before {
        content: '';
        display: block;
        position: absolute;
        top: 100%;
        left: 50%;
        width: 1px;
        height: 100%;
        margin-top: 6px;
        background: ${p.tailColor === 'gray'
          ? p.theme.colors.gray[5]
          : p.theme.colors[p.tailColor]};
      }
    `};

  ${p => p.theme.transition /* sc-declaration */};
`;

const Title = styled.div`
  display: inline-block;
  position: relative;
  align-items: center;
  height: 32px;
  padding-right: 16px;
  color: ${p => {
    switch (p.status) {
      case 'progress':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.error;
      case 'finish':
      case 'wait':
      default:
        return p.theme.colors.gray[5];
    }
  }};
  line-height: 32px;

  ${p =>
    !p.isLast &&
    p.direction === 'horizontal' &&
    css`
      &::after {
        content: '';
        display: block;
        position: absolute;
        top: 16px;
        left: 100%;
        width: 9999px;
        height: 1px;
        background: ${p.tailColor === 'gray'
          ? p.theme.colors.gray[5]
          : p.theme.colors[p.tailColor]};
      }
    `};

  ${p => p.theme.transition /* sc-declaration */};
`;

const Description = styled.div`
  padding-bottom: ${p => p.direction === 'vertical' && '12px'};
  color: ${p => {
    switch (p.status) {
      case 'progress':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.error;
      case 'finish':
      case 'wait':
      default:
        return p.theme.colors.gray[5];
    }
  }};
  font-size: ${p => p.theme.fontSizes.sm};
  ${p => p.theme.transition /* sc-declaration */};
`;

const getStatus = ({ current, count }) => {
  if (current === count) {
    return 'progress';
  }
  if (current > count) {
    return 'finish';
  }
  if (current < count) {
    return 'wait';
  }
};

const getTailColor = ({ count, status, steps }) => {
  if (steps[count + 1] && steps[count + 1].props.status === 'error') {
    return 'error';
  }
  if (status === 'progress' || status === 'wait') {
    return 'gray';
  }
  return 'primary';
};

const StepIcon = ({ status, count, isLast, direction, tailColor, onClick }) => {
  const isFinish = status === 'finish';
  const renderIcon = ['finish', 'error'].includes(status);

  return (
    <StepsIconItem
      status={status}
      isLast={isLast}
      direction={direction}
      tailColor={tailColor}
      onClick={onClick}
    >
      {renderIcon ? (
        <Icon
          size="20"
          type={isFinish ? FinishIcon : ErrorIcon}
          fill={isFinish ? 'primary' : 'error'}
          style={{ pointerEvents: 'none' }}
        />
      ) : (
        count + 1
      )}
    </StepsIconItem>
  );
};

StepIcon.propTypes = {
  count: PropTypes.number.isRequired,
  direction: PropTypes.oneOf(['horizontal', 'vertical']).isRequired,
  isLast: PropTypes.bool.isRequired,
  status: statusType.isRequired,
  tailColor: PropTypes.oneOf(['primary', 'error', 'gray']).isRequired,
  onClick: PropTypes.func.isRequired,
};

export const Step = ({ title, description }) => (
  <Consumer>
    {({ count, status, isLast, tailColor, direction, onCurrentChange }) => (
      <Flex
        flex={isLast ? 'none' : 'auto'}
        mr="16px"
        mb={direction === 'vertical' ? '6px' : 0}
        overflow="hidden"
        minHeight={direction === 'vertical' ? '64px' : 'auto'}
      >
        <StepIcon
          status={status}
          count={count}
          isLast={isLast}
          direction={direction}
          tailColor={tailColor}
          onClick={() => onCurrentChange(count)}
        />
        <div>
          <Title
            status={status}
            direction={direction}
            isLast={isLast}
            tailColor={tailColor}
          >
            {title}
          </Title>
          {description && (
            <Description status={status} direction={direction}>
              {description}
            </Description>
          )}
        </div>
      </Flex>
    )}
  </Consumer>
);

Step.propTypes = {
  /**
   * description of the step, optional property
   */
  description: PropTypes.string,
  /**
   * to specify the status. It will be automatically set by current of Steps
   * if not configured. Optional values are: `wait` `process` `finish` `error`
   */
  // eslint-disable-next-line react/no-unused-prop-types
  status: statusType,
  /**
   * title of the step
   */
  title: PropTypes.string.isRequired,
};

Step.defaultProps = {
  description: '',
  status: null,
};

const Steps = ({ children, current, direction, onCurrentChange }) => {
  const totalCount = Children.count(children) - 1;
  const steps = Children.toArray(children);

  return (
    <Flex flexDirection={direction === 'horizontal' ? 'row' : 'column'}>
      {steps.map((step, count) => {
        const status = step.props.status || getStatus({ current, count });
        const isLast = count === totalCount;
        const tailColor = getTailColor({ count, status, steps });

        return (
          <Provider
            key={step.key}
            value={{
              count,
              status,
              isLast,
              tailColor,
              direction,
              onCurrentChange,
            }}
          >
            {step}
          </Provider>
        );
      })}
    </Flex>
  );
};

Steps.propTypes = {
  /**
   * Step component of Steps.
   */
  children: PropTypes.node.isRequired,
  /**
   * to set the current step, counting from 0. You can overwrite this state by using `status` of `Step`
   */
  current: PropTypes.number.isRequired,
  /**
   * to specify the direction of the step bar, `horizontal` and `vertical` are currently supported
   */
  direction: PropTypes.oneOf(['horizontal', 'vertical']),
  onCurrentChange: PropTypes.func,
};

Steps.defaultProps = {
  direction: 'horizontal',
  onCurrentChange: () => {},
};

Steps.Step = Step;

export default Steps;
