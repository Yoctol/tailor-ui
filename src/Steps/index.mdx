---
name: Steps
menu: Navigation
route: /navigation/steps
---

import { PropsTable, Playground } from 'docz';
import { Value } from 'react-powerplug';

import Steps, { Step } from './';

import Flex from '../Grid/Flex';
import Button from '../Button';
import message from '../message';

# Steps

`Steps` is a navigation bar that guides users through the steps of a task.


## When To Use

When the task is complicated or has a certain sequence in the series of subtasks, we can decompose it into several steps to make things easier.

## Examples

### Basic

<Playground>
  <Steps current={1}>
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br/>
  <Steps current={1}>
    <Steps.Step title="Finished" />
    <Steps.Step title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Error status

<Playground>
  <Steps current={1}>
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step status="error" title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br/>
  <Steps current={1}>
    <Steps.Step title="Finished" />
    <Steps.Step status="error" title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Vertical

<Playground>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br/>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" />
    <Steps.Step title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Vertical & Error status

<Playground>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" description="This is a description" />
    <Steps.Step status="error" title="In Progress" description="This is a description" />
    <Steps.Step title="Waiting" description="This is a description" />
  </Steps>
  <br/>
  <Steps current={1} direction="vertical">
    <Steps.Step title="Finished" />
    <Steps.Step status="error" title="In Progress" />
    <Steps.Step title="Waiting" />
  </Steps>
</Playground>

### Switch Step

<Playground>
  <Value initial={0}>
    {({ value, set }) => {
      const steps = [{
        title: 'First',
        content: 'First-content',
      }, {
        title: 'Second',
        content: 'Second-content',
      }, {
        title: 'Last',
        content: 'Last-content',
      }];

      return (
        <div>
          <Steps current={value} onCurrentChange={set}>
            {steps.map(({ title }) => (
              <Steps.Step key={title} title={title} description="This is a description" />
            ))}
          </Steps>

          <Flex
            my="4"
            height="200px"
            alignItems="center"
            justifyContent="center"
            bg="gray300"
            borderRadius="1"
          >
            {steps[value].content}
          </Flex>

          {value < steps.length - 1 && (
            <Button type="primary" onClick={() => set(value + 1)}>Next</Button>
          )}
          {value === steps.length - 1 && (
            <Button
              type="primary"
              onClick={() => message.success('Processing complete!')}
            >
              Done
            </Button>
          )}
          { value > 0 && (
            <Button ml="2" onClick={() => set(value - 1)}>
              Previous
            </Button>
          )}
        </div>
      );
    }}
  </Value>
</Playground>

## API

### Steps

<PropsTable of={Steps} />

### Steps.Step

<PropsTable of={Step} />
