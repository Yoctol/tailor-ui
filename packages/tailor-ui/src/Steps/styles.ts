import styled, { css } from 'styled-components';

import { Colors, Direction, Status } from './StepContext';

export interface StepsItemProps {
  status: Status;
  direction: Direction;
  isLast: boolean;
  tailColor: Colors | 'gray';
}

export const StepsIconItem = styled.div<StepsItemProps>`
  display: flex;
  position: relative;
  flex: none;
  align-items: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  margin-right: 8px;
  border: ${(p) => p.theme.borders.base};
  border-radius: 50%;
  border-color: ${(p) => {
    switch (p.status) {
      case 'progress':
      case 'finish':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.danger;
      case 'wait':
      default:
        return p.theme.colors.gray500;
    }
  }};
  background-color: ${(p) =>
    p.status === 'progress' ? p.theme.colors.primary : p.theme.colors.light};
  color: ${(p) => {
    switch (p.status) {
      case 'progress':
        return p.theme.colors.light;
      case 'finish':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.danger;
      case 'wait':
      default:
        return p.theme.colors.gray500;
    }
  }};
  cursor: pointer;

  ${(p) =>
    !p.isLast &&
    p.direction === 'vertical' &&
    css`
      &::before {
        content: '';
        display: block;
        position: absolute;
        top: 100%;
        left: 50%;
        width: 1px;
        height: 100%;
        margin-top: 6px;
        background: ${p.tailColor === 'gray'
          ? p.theme.colors.gray500
          : p.theme.colors[p.tailColor]};
      }
    `};

  ${(p) => p.theme.transition /* sc-declaration */};
`;

export const Title = styled.div<StepsItemProps>`
  display: inline-block;
  position: relative;
  align-items: center;
  height: 32px;
  padding-right: 16px;
  color: ${(p) => {
    switch (p.status) {
      case 'progress':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.danger;
      case 'finish':
      case 'wait':
      default:
        return p.theme.colors.gray500;
    }
  }};
  line-height: 32px;

  ${(p) =>
    !p.isLast &&
    p.direction === 'horizontal' &&
    css`
      &::after {
        content: '';
        display: block;
        position: absolute;
        top: 16px;
        left: 100%;
        width: 9999px;
        height: 1px;
        background: ${p.tailColor === 'gray'
          ? p.theme.colors.gray500
          : p.theme.colors[p.tailColor]};
      }
    `};

  ${(p) => p.theme.transition /* sc-declaration */};
`;

export interface DescriptionProps {
  direction: Direction;
  status: Status;
}

export const Description = styled.div<DescriptionProps>`
  padding-bottom: ${(p) => p.direction === 'vertical' && '12px'};
  color: ${(p) => {
    switch (p.status) {
      case 'progress':
        return p.theme.colors.primary;
      case 'error':
        return p.theme.colors.danger;
      case 'finish':
      case 'wait':
      default:
        return p.theme.colors.gray500;
    }
  }};
  font-size: ${(p) => p.theme.fontSizes.sm};
  ${(p) => p.theme.transition /* sc-declaration */};
`;
