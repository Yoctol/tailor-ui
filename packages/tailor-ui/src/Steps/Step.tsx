import React, { FC, useContext } from 'react';
import { MdClose, MdDone } from 'react-icons/md';

import { Flex } from '../Layout';
import { Icon } from '../Icon';

import StepContext, { Status } from './StepContext';
import { Description, StepsIconItem, StepsItemProps, Title } from './styles';

type StepIconProps = StepsItemProps & {
  count: number;
  onClick: () => void;
};

const StepIcon: FC<StepIconProps> = ({
  status,
  count,
  isLast,
  direction,
  tailColor,
  onClick,
}) => {
  const isFinish = status === 'finish';
  const renderIcon = ['finish', 'error'].includes(status);

  return (
    <StepsIconItem
      status={status}
      isLast={isLast}
      direction={direction}
      tailColor={tailColor}
      onClick={onClick}
    >
      {renderIcon ? (
        <Icon
          size="20"
          type={isFinish ? MdDone : MdClose}
          fill={isFinish ? 'primary' : 'error'}
          style={{ pointerEvents: 'none' }}
        />
      ) : (
        count + 1
      )}
    </StepsIconItem>
  );
};

export interface StepProps {
  title: string;
  description?: string;
  status?: Status;
}

const Step: FC<StepProps> = ({ title, description = null }) => {
  const {
    count,
    status,
    isLast,
    tailColor,
    direction,
    onCurrentChange,
  } = useContext(StepContext);
  return (
    <Flex
      flex={isLast ? 'none' : 'auto'}
      mr="16px"
      mb={direction === 'vertical' ? '6px' : 0}
      overflow="hidden"
      minHeight={direction === 'vertical' ? '64px' : 'auto'}
    >
      <StepIcon
        status={status}
        count={count}
        isLast={isLast}
        direction={direction}
        tailColor={tailColor}
        onClick={() => {
          if (onCurrentChange) {
            onCurrentChange(count);
          }
        }}
      />
      <div>
        <Title
          status={status}
          direction={direction}
          isLast={isLast}
          tailColor={tailColor}
        >
          {title}
        </Title>
        {description && (
          <Description status={status} direction={direction}>
            {description}
          </Description>
        )}
      </div>
    </Flex>
  );
};

export default Step;
