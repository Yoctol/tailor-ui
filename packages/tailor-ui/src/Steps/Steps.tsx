import React, {
  Children,
  FC,
  ReactElement,
  ReactNode,
  isValidElement,
} from 'react';

import { Flex } from '../Layout';

import Step, { StepProps } from './Step';
import StepContext, { Direction, Status } from './StepContext';

const getStatus = ({ current, count }: { current: number; count: number }) => {
  if (current === count) {
    return 'progress';
  }
  if (current > count) {
    return 'finish';
  }
  if (current < count) {
    return 'wait';
  }

  return 'progress';
};

const getTailColor = ({
  count,
  status,
  steps,
}: {
  count: number;
  status: Status;
  steps: ReactNode[];
}) => {
  const nextStep = steps[count + 1];
  if (
    isValidElement<StepProps>(nextStep) &&
    nextStep.props.status === 'error'
  ) {
    return 'error';
  }

  if (status === 'progress' || status === 'wait') {
    return 'gray';
  }

  return 'primary';
};

export interface StepsProps {
  current: number;
  direction?: Direction;
  onCurrentChange?: () => void;
  children: ReactElement<StepProps>[];
}

const Steps: FC<StepsProps> & {
  Step: typeof Step;
} = ({
  children,
  current,
  onCurrentChange,
  direction = 'horizontal' as Direction,
}) => {
  const totalCount = Children.count(children) - 1;
  const steps = Children.toArray(children);

  return (
    <Flex flexDirection={direction === 'horizontal' ? 'row' : 'column'}>
      {steps.map((step, count) => {
        if (isValidElement<StepProps>(step)) {
          const status = step.props.status || getStatus({ current, count });
          const isLast = count === totalCount;
          const tailColor = getTailColor({ count, status, steps });

          return (
            <StepContext.Provider
              key={step.key || count}
              value={{
                count,
                status,
                isLast,
                tailColor,
                direction,
                onCurrentChange,
              }}
            >
              {step}
            </StepContext.Provider>
          );
        }

        return null;
      })}
    </Flex>
  );
};

Steps.Step = Step;

export { Steps };
