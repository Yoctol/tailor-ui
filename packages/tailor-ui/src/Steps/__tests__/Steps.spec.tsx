import React from 'react';

import { render } from 'test/test-utils';

import { Steps } from '../Steps';

describe('Steps', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Steps current={1}>
        <Steps.Step title="Finished" description="This is a description" />
        <Steps.Step
          status="error"
          title="In Progress"
          description="This is a description"
        />
        <Steps.Step title="Waiting" description="This is a description" />
      </Steps>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
