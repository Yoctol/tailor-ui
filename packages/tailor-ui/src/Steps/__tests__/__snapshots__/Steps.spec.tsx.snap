// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  overflow: hidden;
  margin-bottom: 0rem;
  margin-right: 16px;
  min-height: auto;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  overflow: hidden;
  margin-bottom: 0rem;
  margin-right: 16px;
  min-height: auto;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
}

.c3 svg {
  fill: #8b8b8b;
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c3.c3.c3 svg {
  fill: #3c5ad0;
  width: 20px;
  height: 20px;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
}

.c7 svg {
  fill: #8b8b8b;
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c7.c7.c7 svg {
  fill: #fc5475;
  width: 20px;
  height: 20px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  margin-right: 8px;
  border: 1px solid;
  border-radius: 50%;
  border-color: #3c5ad0;
  background-color: #ffffff;
  color: #3c5ad0;
  cursor: pointer;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  margin-right: 8px;
  border: 1px solid;
  border-radius: 50%;
  border-color: #fc5475;
  background-color: #ffffff;
  color: #fc5475;
  cursor: pointer;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  margin-right: 8px;
  border: 1px solid;
  border-radius: 50%;
  border-color: #8b8b8b;
  background-color: #ffffff;
  color: #8b8b8b;
  cursor: pointer;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c4 {
  display: inline-block;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 32px;
  padding-right: 16px;
  color: #8b8b8b;
  line-height: 32px;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c4::after {
  content: '';
  display: block;
  position: absolute;
  top: 16px;
  left: 100%;
  width: 9999px;
  height: 1px;
  background: #fc5475;
}

.c8 {
  display: inline-block;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 32px;
  padding-right: 16px;
  color: #fc5475;
  line-height: 32px;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c8::after {
  content: '';
  display: block;
  position: absolute;
  top: 16px;
  left: 100%;
  width: 9999px;
  height: 1px;
  background: #3c5ad0;
}

.c12 {
  display: inline-block;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 32px;
  padding-right: 16px;
  color: #8b8b8b;
  line-height: 32px;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c5 {
  color: #8b8b8b;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c9 {
  color: #fc5475;
  font-size: 0.875rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

<div
  class="c0"
>
  <div
    class="c1"
    overflow="hidden"
  >
    <div
      class="c2"
      direction="horizontal"
    >
      <i
        class="c3"
        cursor="default"
        fill="primary"
        size="20"
        style="pointer-events: none;"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </i>
    </div>
    <div>
      <div
        class="c4"
        direction="horizontal"
      >
        Finished
      </div>
      <div
        class="c5"
        direction="horizontal"
      >
        This is a description
      </div>
    </div>
  </div>
  <div
    class="c1"
    overflow="hidden"
  >
    <div
      class="c6"
      direction="horizontal"
    >
      <i
        class="c7"
        cursor="default"
        fill="error"
        size="20"
        style="pointer-events: none;"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </i>
    </div>
    <div>
      <div
        class="c8"
        direction="horizontal"
      >
        In Progress
      </div>
      <div
        class="c9"
        direction="horizontal"
      >
        This is a description
      </div>
    </div>
  </div>
  <div
    class="c10"
    overflow="hidden"
  >
    <div
      class="c11"
      direction="horizontal"
    >
      3
    </div>
    <div>
      <div
        class="c12"
        direction="horizontal"
      >
        Waiting
      </div>
      <div
        class="c5"
        direction="horizontal"
      >
        This is a description
      </div>
    </div>
  </div>
</div>
`;
